src/lib/core/cbus.h

```diff
@@ -466,6 +488,66 @@ cbus_unpair(struct cpipe *dest_pipe, struct cpipe *src_pipe,
 	    void (*unpair_cb)(void *), void *unpair_arg,
 	    void (*process_cb)(struct cbus_endpoint *));
 
+/**
+ * A  uni-directional FIFO queue from any thread to cord.
+ * It is a light version of the cpipe, since some features are not needed for
+ * 3td party threads or they cannot be implemented.
+ *
+ */
+struct lcpipe {
+	/** Staging area for pushed messages */
+	struct stailq input;
+	/** Counters are useful for finer-grained scheduling. */
+	int n_input;
+	/**
+	 * When pushing messages, keep the staged input size under
+	 * this limit (speeds up message delivery and reduces
+	 * latency, while still keeping the bus mutex cold enough).
+	 */
+	int max_input;
+	/**
+	 * The cbus endpoint at the destination cord to handle
+	 * flushed messages.
+	 */
+	struct cbus_endpoint *endpoint;
+};
+
```

Comment:
Given that this structure itself and all its methods are a complete copy-paste of some parts of `cpipe`, I would suggest to find a way to reuse the code.

One suggestion could be that you do the C-inheritance. So this struct must be added into `struct cpipe` as `base` member instead of those duplicated members here. Then you make the relevant `cpipe` methods call the `lcpipe` methods on its `&base` member pointer.

If my guess about `ev_invoke` behaviour above is correct, then this should be a simple change which will eliminate a lot of duplicated code.

Reply:
Done, but I couldn’t ifx code duplication this way (except in cbus.h), because wherever cpipe is used, both its base part and other fields are required

Reply:
The main idea was exactly to fix the duplication in cbus.c. Please, read a bit about C struct inheritance and check some examples maybe in how `struct port`. Applied to cbus code this is what I've meant:
<details><summary>Diff</summary>

```Diff
diff --git a/src/lib/core/cbus.c b/src/lib/core/cbus.c
index 2ada0eef9..0cc9b17e2 100644
--- a/src/lib/core/cbus.c
+++ b/src/lib/core/cbus.c
@@ -139,36 +139,33 @@ cbus_endpoint_poison_f(struct cmsg *msg)
 	free(msg);
 }
 
-void
-cpipe_destroy(struct cpipe *pipe)
+static void
+lcpipe_destroy(struct lcpipe *pipe)
 {
 	/*
 	 * The thread should not be canceled while mutex is locked.
 	 * And everything else must be protected for consistency.
-	*/
+	 */
 	int old_cancel_state;
 	tt_pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &old_cancel_state);
 
-	ev_async_stop(pipe->producer, &pipe->flush_input);
-
 	static const struct cmsg_hop route[1] = {
 		{cbus_endpoint_poison_f, NULL}
 	};
-	trigger_destroy(&pipe->on_flush);
 
-	struct cbus_endpoint *endpoint = pipe->base.endpoint;
-	struct cmsg_poison *poison = malloc(sizeof(struct cmsg_poison));
+	struct cbus_endpoint *endpoint = pipe->endpoint;
+	struct cmsg_poison *poison = xmalloc(sizeof(*poison));
 	cmsg_init(&poison->msg, route);
-	poison->endpoint = pipe->base.endpoint;
+	poison->endpoint = pipe->endpoint;
 	/*
-	 * Avoid the general purpose cpipe_push_input() since
+	 * Avoid the general purpose lcpipe_push() since
 	 * we want to control the way the poison message is
 	 * delivered.
 	 */
 	tt_pthread_mutex_lock(&endpoint->mutex);
 	/* Flush input */
-	stailq_concat(&endpoint->output, &pipe->base.input);
-	pipe->base.n_input = 0;
+	stailq_concat(&endpoint->output, &pipe->input);
+	pipe->n_input = 0;
 	/* Add the pipe shutdown message as the last one. */
 	stailq_add_tail_entry(&endpoint->output, poison, msg.fifo);
 	/* Count statistics */
@@ -183,7 +180,21 @@ cpipe_destroy(struct cpipe *pipe)
 	tt_pthread_mutex_unlock(&endpoint->mutex);
 
 	tt_pthread_setcancelstate(old_cancel_state, NULL);
+}
+
+void
+lcpipe_delete(struct lcpipe *pipe)
+{
+	lcpipe_destroy(pipe);
+	free(pipe);
+}
 
+void
+cpipe_destroy(struct cpipe *pipe)
+{
+	ev_async_stop(pipe->producer, &pipe->flush_input);
+	trigger_destroy(&pipe->on_flush);
+	lcpipe_destroy(&pipe->base);
 	TRASH(pipe);
 }
 
@@ -275,50 +286,7 @@ lcpipe_push_now(struct lcpipe *pipe, struct cmsg *msg)
 	lcpipe_flush_input(pipe);
 }
 
-void
-lcpipe_delete(struct lcpipe *pipe)
-{
-	/*
-	 * The thread should not be canceled while mutex is locked.
-	 * And everything else must be protected for consistency.
-	 */
-	int old_cancel_state;
-	tt_pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &old_cancel_state);
-
-	static const struct cmsg_hop route[1] = {
-		{cbus_endpoint_poison_f, NULL}
-	};
-
-	struct cbus_endpoint *endpoint = pipe->base.endpoint;
-	struct cmsg_poison *poison = xmalloc(sizeof(*poison));
-	cmsg_init(&poison->msg, route);
-	poison->endpoint = pipe->base.endpoint;
-	/*
-	 * Avoid the general purpose lcpipe_push() since
-	 * we want to control the way the poison message is
-	 * delivered.
-	 */
-	tt_pthread_mutex_lock(&endpoint->mutex);
-	/* Flush input */
-	stailq_concat(&endpoint->output, &pipe->base.input);
-	pipe->base.n_input = 0;
-	/* Add the pipe shutdown message as the last one. */
-	stailq_add_tail_entry(&endpoint->output, poison, msg.fifo);
-	/* Count statistics */
-	rmean_collect(cbus.stats, CBUS_STAT_EVENTS, 1);
-	/*
-	 * Keep the lock for the duration of ev_async_send():
-	 * this will avoid a race condition between
-	 * ev_async_send() and execution of the poison
-	 * message, after which the endpoint may disappear.
-	 */
-	ev_async_send(endpoint->consumer, &endpoint->async);
-	tt_pthread_mutex_unlock(&endpoint->mutex);
-
-	tt_pthread_setcancelstate(old_cancel_state, NULL);
 
-	free(pipe);
-}
 
 static void
 cbus_create(struct cbus *bus)
diff --git a/src/lib/core/cbus.h b/src/lib/core/cbus.h
index c4e48d850..cdb20c8df 100644
--- a/src/lib/core/cbus.h
+++ b/src/lib/core/cbus.h
@@ -110,9 +110,12 @@ void
 cmsg_deliver(struct cmsg *msg);
 
 /**
- * A common part of `cpipe` and `lcpipe`.
+ * A  uni-directional FIFO queue from any thread to cord.
+ * It is a light version of the cpipe, since some features are not needed for
+ * 3td party threads or they cannot be implemented.
+ *
  */
-struct pipe_base {
+struct lcpipe {
 	/** Staging area for pushed messages */
 	struct stailq input;
 	/** Counters are useful for finer-grained scheduling. */
@@ -133,7 +136,7 @@ struct pipe_base {
 /** A  uni-directional FIFO queue from one cord to another. */
 struct cpipe {
 	/** Base part of a pipe. */
-	struct pipe_base base;
+	struct lcpipe base;
 	/**
 	 * Rather than flushing input into the pipe
 	 * whenever a single message or a batch is
@@ -496,17 +499,6 @@ cbus_unpair(struct cpipe *dest_pipe, struct cpipe *src_pipe,
 	    void (*unpair_cb)(void *), void *unpair_arg,
 	    void (*process_cb)(struct cbus_endpoint *));
 
-/**
- * A  uni-directional FIFO queue from any thread to cord.
- * It is a light version of the cpipe, since some features are not needed for
- * 3td party threads or they cannot be implemented.
- *
- */
-struct lcpipe {
-	/** Base part of a pipe. */
-	struct pipe_base base;
-};
-
 /**
  * Create and initialize a pipe and connect it to the consumer.
  * The call returns only when the consumer, identified by consumer name, has
```
</details>

Please, rework the code in this way. lcpipe is the base, and cpipe reuses it similar to how I show it in the diff above.